lib/webid_util.js

```diff
@@ -0,0 +1,266 @@
+          'Accept': 'text/turtle, application/ld+json'
+        }
+      }
+
+      request(options, function (err, res, body) {
+        if (err) {
+          return reject(new Error('Failed to fetch profile from ' + uri.href + ': ' + err))
+        }
+
+        if (res.statusCode !== 200) {
+          return reject(new Error('Failed to retrieve WebID from ' + uri.href + ': HTTP status: ' + res.statusCode))
+        }
+
+        resolve({profile: body, contentType: res.headers['content-type']})
+      })
+    })
+  },
+
+  statementExists: function (kb, subject, predicate, object) {
+    var s = kb.sym(subject)
+    var p = kb.sym(predicate)
+    var o = kb.sym(object)
+    var st = kb.statementsMatching(s, p, o)
+    return st.length > 0
+  }, // statementExists
+
+  findDelegatePublicKeyInDelegatorProfile: function (kb, delegate, delegateModulus, delegateExponent) {
+    return new Promise(function (resolve, reject) {
+      var rc = []
+      delegateModulus = delegateModulus.toLowerCase()
+      delegateExponent = delegateExponent.toLowerCase()
+      var sparql = `
+        PREFIX cert: <http://www.w3.org/ns/auth/cert#> 
+        SELECT ?exponent ?modulus
+        WHERE {
+          <${delegate}> cert:key ?key .
+          ?key cert:exponent ?exponent .
+          ?key cert:modulus ?modulus .
+        }
+        `
+
+      var onresult = function (result) {
+        if (result) {
+          var exponent = result['?exponent'].value
+          var modulus = result['?modulus'].value
+          rc.push({exponent, modulus})
+        }
+      }
+
+      var onDone = function () {
+        var i
+        var publicKey = null
+
+        for (i = 0; i < rc.length; i++) {
+          if (rc[i].modulus.toLowerCase() === delegateModulus &&
+              parseInt(rc[i].exponent) === parseInt(delegateExponent)) {
+            publicKey = { exponent: delegateExponent, modulus: delegateModulus }
+          }
+        }
+        resolve(publicKey)
+      }
+
+      const query = $rdf.SPARQLToQuery(sparql, false, kb)
+      kb.query(query, onresult, undefined, onDone)
+    })
+  },
+
+  findInfoInWebIdProfile: function (kb, webId) {
+    var rc
+    var name = null
+    var webIdUri = kb.sym(webId)
+
+    rc = kb.any(webIdUri, kb.sym('http://schema.org/name'))
+    if (rc) {
+      name = rc.value
+    } else {
+      rc = kb.any(webIdUri, kb.sym('http://xmlns.com/foaf/0.1/name'))
+      if (rc) {
+        name = rc.value
+      } else {
+        rc = kb.any(webIdUri, kb.sym('http://www.w3.org/2000/01/rdf-schema#label'))
+        if (rc) {
+          name = rc.value
+        } else {
+          rc = kb.any(webIdUri, kb.sym('http://www.w3.org/2004/02/skos/core#prefLabel'))
+          if (rc) {
+            name = rc.value
+          } else {
+            rc = kb.any(webIdUri, kb.sym('http://www.w3.org/2004/02/skos/core#altLabel'))
+            if (rc) {
+              name = rc.value
+            }
+          }
+        }
+      }
+    }
+    return {name}
```

Comment:
We should avoid this pyramid code, that's hard to maintain. We could for instance put all those predicates in an array and do a reduce.

Reply:
Fixed